io/shared/src/main/scala/scalaz/effect/IO.scala

```diff
@@ -486,6 +458,47 @@ sealed abstract class IO[E, A] { self =>
 }
 
 object IO {
+
+  trait Disj[A, B] {
+    def fold[Z](left: A => Z, right: B => Z): Z
+  }
+
+  object Disj {
+    def either[A, B](e: Either[A, B]): Disj[A, B] =
+      new Disj[A, B] {
+        def fold[Z](left: A => Z, right: B => Z): Z =
+          e.fold(left, right)
+      }
+    def option[B](o: Option[B]): Disj[Unit, B] =
+      new Disj[Unit, B] {
```

Comment:
This should use a preallocated object, cast to the right type.

Reply:
Ditto.